## Project - Diabetes Prediction Using Machine Learning

## Problem Statement

Early detection of diabetes plays a critical role in preventing long-term complications. However, many individuals remain undiagnosed due to limited screening. This project aims to develop a machine learning model that predicts whether an individual is diabetic based on key medical and lifestyle indicators.

## Objective

- Analyze medical features to understand factors influencing diabetes.

- Build a reliable classification model to predict diabetes status.

- Evaluate model performance using accuracy and other metrics.

- Enable early risk identification using a simple, data-driven ML approach.

In [1]:
#importing the dependencies 

import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler 
from sklearn.model_selection import train_test_split 
from sklearn import svm
from sklearn.metrics import accuracy_score

### Data Collection & Analysis 

In [2]:
#loading the diabetes dataset to a pandas data frame 

df = pd.read_csv("diabetes.csv",lineterminator="\n")
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome\r  
0                       0.627   50          1  
1                       0.351   31          0  
2                       0.672   32          1  
3                       0.167   21          0  
4                       2.288   33          1  
..                        ...  ...        ...  
763                     0.171   63          0  
764                     0.340   27          0  
765                     0.245   30          0  
766                     0.349   47          1  
767                     0.315   23          0  

[768 rows x 9 columns]

**Dataset Information**

##### The dataset contains 768 records with the following features:

- Pregnancies – number of times the individual has been pregnant
- Glucose – plasma glucose concentration during an oral glucose tolerance test
- BloodPressure – diastolic blood pressure (mm Hg)
- SkinThickness – triceps skinfold thickness (mm)
- Insulin – 2-hour serum insulin (mu U/ml)
- BMI – Body Mass Index (weight in kg / height in m²)
- DiabetesPedigreeFunction – likelihood of diabetes based on family history
- Age – age of the individual
- Outcome – target variable (1 = diabetic, 0 = non-diabetic)

In [3]:
#first 5 rows of the dataset

df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome\r  
0                     0.627   50          1  
1                     0.351   31          0  
2                     0.672   32          1  
3                     0.167   21          0  
4                     2.288   33          1

In [4]:
#checking the number of rows and columns 

df.shape

(768, 9)

In [5]:
#getting the statistical summary of the numerical data 

df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age   Outcome\r  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [6]:
#getting some informations about the dataset 

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
                  768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [7]:
#removing all the unwanted things from the column names

df.columns = df.columns.str.strip()
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [8]:
#checking the diabetics status of the patients

df["Outcome"].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

#### 0 --> Non-diabetic 
#### 1 --> Diabetic 

In [9]:
#displays avg of the dataset based on Diabetic and Non-diabetic patients 

df.groupby("Outcome").mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

In [10]:
#checking for missing values in the dataset 

df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [11]:
#checking for duplicate values in the dataset 

df.duplicated().sum()

np.int64(0)

### Splitting the Features and Target

In [12]:
#separating the data and labels

X = df.drop(columns = "Outcome", axis = 1)
Y = df["Outcome"]

In [13]:
print(X)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [14]:
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


### Data Standardization 

In [15]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [16]:
print(X_scaled)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [17]:
#saving the variable named X 

X = X_scaled

### Splitting the data into Training and Testing Part

In [18]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=42)

In [19]:
print(X.shape,X_train.shape,X_test.shape)

(768, 8) (614, 8) (154, 8)


### Training the Model

In [20]:
model = svm.SVC(kernel="linear")

In [21]:
#training the support vector machine classifier 

model.fit(X_train,Y_train)

SVC(kernel='linear')

### Model Evaluation 

In [22]:
#accuracy score on the training data 

X_train_pred = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_pred,Y_train)

In [23]:
print("Training data accuracy score: ",training_data_accuracy)

Training data accuracy score:  0.7719869706840391


In [24]:
#accuracy score on the testing data 

X_test_pred = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_pred,Y_test)

In [25]:
print("Testing data accuracy score: ",testing_data_accuracy)

Testing data accuracy score:  0.7597402597402597


### Making a Predictive System 

In [27]:
input_data = (4,129,86,20,270,35.1,0.231,23)

#changing the input data to numpy array
input_data_numpy_array = np.asarray(input_data)

#reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_numpy_array.reshape(1,-1)

#standardize the input data 
std_data = scaler.transform(input_data_reshaped)
print(std_data)

prediction = model.predict(std_data)
print(prediction)


if (prediction[0] == 1):
    print("The person is Diabetic.")
else:
    print("The person is not Diabetic.")

[[ 0.04601433  0.25367803  0.87340925 -0.03365099  1.65149133  0.39439164
  -0.72747502 -0.87137393]]
[0]
The person is not Diabetic.


C:\ProgramData\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


### Conclusion 

**Key Insights**

- Glucose and BMI show the strongest association with diabetes.
- Older individuals tend to have a higher chance of being diabetic.
- Zero values in medical features must be treated before training the model.

**Final Takeaway**

This ML project demonstrates how basic health features can be transformed into a predictive system that assists with early diabetes detection, supporting preventive healthcare efforts.